### PR TITLE
TASK-17: Event Page

### PR DESCRIPTION
**Pull request by cristian-aldea**
This PR addressed #70 

This PR aims to implement an event page for use in other user stories.

The images used in the component are placeholders until saving to and querying from the media server is done.

You can access the page by going to http://localhost/test-event-page

Notes:
* Added interfaces for events and related entities, and made the existing entities inherit those interfaces. That way the front-end can use these interfaces.
* Adjusted styling of html, body, and container divs to take up the entire width and height of the page. Without this, the scrolling in the event page info would be much less feasible. 
* Fixed useless test in apps/mull-ui/src/app/components/header/header.spec.tsx
* Moved button from create event page into its own component, under a components/ folder
* created EventRestrictionMap for easy access to enum string values
* Fixed inter font not being loaded due to incorrect url
* although this issue is blocked by #78, I created the PR to get feedback

### CODE REVIEW

**Review by r-saba — CHANGES_REQUESTED** [review]
Looking great! Here's my UI comments

- The green horizontal bar between event info fields is too dark compared to figma
- Missing the last green horizontal bar, there's one under each info row
- The icons on the left are too large
- The event description should have "view more" and once we click that we see the entire text
- Missing the last row, "invite friends"

**Comment by cristian-aldea** [comment]
Just a note @r-saba, please try to not have too many comments in the main review text, I did that mistake before and it makes it hard to reply (since I can only quote reply)

Here is my response: 
> * The green horizontal bar between event info fields is too dark compared to figma

Sure thing will fix

> * Missing the last green horizontal bar, there's one under each info row

If you look at the mockup you'll see that the divider lines are a bit inconsistent
image
I tried seeing how it would look if I added lines for the top of the first row and the bottom of the last row, but I didn't like how it looked. We can discuss the specific styling and what we should go for in a meeting

> * The icons on the left are too large

True, ill fix that

> * The event description should have "view more" and once we click that we see the entire text

Ritchelle and I discussed and we felt that that might not be necessary, I can implement it real quick if we need its needed.

> * Missing the last row, "invite friends"

True, didn't include it at the time since a few details we're still shaky, will add it.

**Comment by r-saba** [comment]
@TheGreatMarkus Sure will make single comments next time.
Regarding the missing divider at the bottom, do whatever you feel looks nice.
Regarding the "view more", I do think it's necessary. Unless we limit the number of characters in the event description, it could end up taking too much space on this page. If you could implement real quick that'd be sweet thanks!

**Comment by cristian-aldea** [comment]
@r-saba sounds good, I'll implement the read more thing, and as for the dividers I can ask that real quick during our scrum

**Comment by cristian-aldea** [comment]
@r-saba implemented the expandable text as a separate component, I'm quite happy with how it turned, since its very flexible and easy to use imo. Lemme know what you think, I've addressed all your comments (aside from the div lines, which we'll talk during the meeting today), so I've asked for your review again.

**Comment by RGPosadas** [comment]
Looked over the UI and I don't have any comments other than good job, it looks and works great! I am just waiting for #78 to be merged so that I can give you a full review.

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Matches better the mockup and the mockups use 0.5 for width.
```suggestion
    border-bottom: 0.5px solid rgba($color: $primary-color, $alpha: 0.5);
```

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Title is not bolded in the mockup. 

ps - Since this is on a h1, and we haven't cleared the styles the `font-weight: bold` is unnecessary since it's bold by default styling.

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: We're not using flex, by default it will take up the entire space.
```suggestion
```

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Is there a reason for this?

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Is there a reason for this?

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Typically we use max-width when we want something to stay below 100%. `max-width: 50%`.
This isn't necessary here.
```suggestion
```

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: I don't typically use `border-box` why is this everywhere? When I remove it in the inspector nothing changes

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: - `flex-direction` is row by default not need to explicitly define it.
- row will be width 100% by default no need to explicitly define it.
- Do we need `align-items: flex-start`? I don't usually have to use this.

File: libs/ui-lib/src/components/mull-button/mull-button.scss
Comment (r-saba) [inline libs/ui-lib/src/components/mull-button/mull-button.scss]: Button doesn't have a border by default anyway in our app.
```suggestion
.mull-button {
  color: white;
  font-weight: bold;
  padding: 1rem 3rem;
  background: $primary-color;
  border-radius: 9px;
```

File: libs/ui-lib/src/components/expandable-text/expandable-text.tsx
Comment (r-saba) [inline libs/ui-lib/src/components/expandable-text/expandable-text.tsx]: I would move this into a function and call the function here. Typically we don't want "complex" logic inside the return statement.

File: libs/ui-lib/src/lib/styles/styles.scss
Comment (r-saba) [inline libs/ui-lib/src/lib/styles/styles.scss]: I don't like this, have you considered how it will affect other styling? 
I see you're doing this to keep the button at the bottom I know there was a discussion on if we're keeping it sticky or part of the scrolling section. If it needs to stick to the bottom, consider position abolsute.

**Review by r-saba — CHANGES_REQUESTED** [review]
Since the PR is strictly front-end I'm paying a lot more attention to the CSS. Just be careful with what you add. CSS that has not been properly thought through can cause a lot of head ache for the next developer when their elements don't style as expected. Also unnecessary/dead CSS is confusing. When in doubt steer towards specificity.

We typically don't have to specify height/width 100% unless we're using something like flex or some other styling is affecting our component.

- Let's replace `em` with `rem` to be consistent. I noticed you were using both in this pr.
- Double check all your CSS

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: You're right, ill make it a div and remove the font weight

**Comment by cristian-aldea** [comment]
thanks @r-saba for giving this so much attention, will look into your comments. Just had some quick replies to your comments:

> * Let's replace `em` with `rem` to be consistent. I noticed you were using both in this pr.

We can both use `em` and `rem` when suitable, no needed to force ourselves to use one over the other. I do admit some of my uses of `em` were accidental/bad, but just felt like pointing that out.

> * Double check all your CSS

will do o7

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: I set a max height for the text of 3 lines (3em) to account for long titles. If the title ends up longer than that I make it scrollable.

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: See above

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Removing this

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: Removing this

Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: I like to use box-sizing: border-box when I center since otherwise things get misaligned sometimes due to padding. Not necessary in this case.

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Thanks for pointing that out, I'm not familiar with all the default, will be removing these lines

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Yup, will also simplify this, since we don't need to exclude the last children from having a border

File: libs/ui-lib/src/components/expandable-text/expandable-text.tsx
Comment (cristian-aldea) [reply libs/ui-lib/src/components/expandable-text/expandable-text.tsx]: You would move the tsx here to a function? I don't feel like that's necessary. lets discuss about this together

File: libs/ui-lib/src/components/mull-button/mull-button.scss
Comment (cristian-aldea) [reply libs/ui-lib/src/components/mull-button/mull-button.scss]: I think this was left over due to jimmy and jon putting a border for some reason on the original button I moved out of create event.

I probably set it to none just in case, even though its not needed, much like the rest of the dead css in the rest of this pr haha

File: libs/ui-lib/src/lib/styles/styles.scss
Comment (cristian-aldea) [reply libs/ui-lib/src/lib/styles/styles.scss]: This isn't to have the button stuck at the bottom. This is to make the event info component scroll correctly doable. For the event info to scroll but not the event page itself, the event page needs to take up the whole page, which wasn't possible since `html`, `body` and `#root` had `fit-content` as their height.

I really doubt this will negatively affect any styling imo, since all it does is make sure that containers take up the entire page, which I thought it was doing already. Also I tested the existing page and no stylng broke as far as I know. 

This warrants a longer conversation, thought, and obviously I might be wrong on what I said earlier. Open to alternatives since you have more experience than me on this.

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: As we discussed, we will assume title aren't absurdly long

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]: caused by the width:100% in app.scss. Removing all/most of my uses of border-box since its no longer necessary

File: libs/ui-lib/src/lib/styles/styles.scss
Comment (cristian-aldea) [reply libs/ui-lib/src/lib/styles/styles.scss]: found an alternative solution. The height of the page is 100vh - height of navbar - height of header

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: I find the rows could use more breathing room. The spacing in the mockups is not consistent so it's hard to say exactly what we should use. I suggest 1.25rem.
```suggestion
  padding: 1.25rem 1rem;
```

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Button could use more margin-top to match the mockup and look better.

**Review by r-saba — CHANGES_REQUESTED** [review]
Thanks for addressing all my comments. Really good work! I left just a very minor feedback on the UI below.

File: libs/ui-lib/src/lib/styles/styles.scss
Comment (r-saba) [reply libs/ui-lib/src/lib/styles/styles.scss]: Nice! That's what I was getting at with `vh` during our discussion, very useful!

**Comment by r-saba** [comment]
Regarding `em` vs `rem`. I haven't seen the usage of `em` at all actually in web apps. rem is a standard and I suggest we stick with it unless we really have a specific use case to use `em`. Using both will result in inconsistencies throughout our app.

**Comment by cristian-aldea** [comment]
> Regarding `em` vs `rem`. I haven't seen the usage of `em` at all actually in web apps. rem is a standard and I suggest we stick with it unless we really have a specific use case to use `em`. Using both will result in inconsistencies throughout our app.

I disagree, from reading articles like https://zellwk.com/blog/rem-vs-em/, each have their usecase. `em` is also a standard, just like `rem`. It's true that so far we haven't had much of a usecase for `em` in our code, but I just don't want the use of `rem` to be dogmatic.

For example, if we end up needing to define bottom padding for different font sizes (p, h1, h2, h3, etc.) that would be a perfect use case for em, since we'd want the padding to change depending on the size of the current element's text.

**Comment by cristian-aldea** [comment]
Sorry for all the pushes @r-saba, i tried by best to clean up the git history

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Nice catch, I do agree that some more padding is needed, however I feel that 1.25 is excessive from my testing. I bumped up the padding to 1rem:

implementation:
image

mockup:
image

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Yup agree, top and bottom should have more margin

**Comment by r-saba** [comment]
As I mentioned I haven't seen `em` used. Even the website you linked is using rem. Also I've found [plenty ](https://webdesign.tutsplus.com/tutorials/comprehensive-guide-when-to-use-em-vs-rem--cms-23984) of articles suggesting exactly what I said, the usage of `rem` since it's always consistent and using `em` only if it is necessary. 

I haven't worked with design systems that spaces content based on font size as you suggest. Typically we have a defined spacing for layout (40px b/w headings and content, 10px between content and buttons etc..). We're saying the same thing that `em` has some use cases in sepcific situations (though I've looked the css of many website over the years and haven't seen it used). However we should use `rem` where we can since it's behavior is consistent and easier to predict, and use `em` only if we find a good use case for it. Using both liberally will make our UI look unprofessional.

File: apps/mull-ui/src/app/components/mull-back-button/mull-back-button.tsx
Comment (r-saba) [inline apps/mull-ui/src/app/components/mull-back-button/mull-back-button.tsx]:
```suggestion
      onClick={onClick || defaultOnClick}
```

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.scss]:
```suggestion

```

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx
Comment (r-saba) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx]: can you rename to `prevPage`? `lastPage` sounds like it's related to pagination

File: apps/mull-ui/src/app/components/mull-button/mull-button.scss
Comment (r-saba) [inline apps/mull-ui/src/app/components/mull-button/mull-button.scss]: Unnecessary spacing
```suggestion
```

File: apps/mull-ui/src/app/pages/event-page/event-page.tsx
Comment (r-saba) [inline apps/mull-ui/src/app/pages/event-page/event-page.tsx]: If you change it to `prevPage` above, change it here too

**Review by r-saba — DISMISSED** [review]
LGTM! None of my comments below are blockers. If you do address these comments let me know. If you tell me you didn't touch any other part of the code I will approve it quickly after testing functionality.

File: apps/mull-ui/src/app/components/mull-back-button/mull-back-button.tsx
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/mull-back-button/mull-back-button.tsx]: I had a feeling there was a better way of doing this. Thanks!

**Comment by cristian-aldea** [comment]
@r-saba Yeah we're basically saying the same thing, and I admit my uses of `em` in earlier sprints and in this pr were mostly due to ignorance or error. Just wanted to make sure people reading this understand why we use one over the other.

**Review by RGPosadas — DISMISSED** [review]
**Subtasks**
- [x] 17.2 Event page header
- [x] 17.3 Event info
- [x] 17.4 Mull button
- [x] 17.5 Integrate Components into this task.

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=2007%3A404529)

**Comments**
Great job! The added padding between the div and the row content (compared to the first time I reviewed this PR) is a great subtle addition.
LGTM!

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx
Comment (JimLau49) [inline apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx]: I think we can have svg imports without spaces in between them

**Review by JimLau49 — DISMISSED** [review]
LGTM! , really nice work you have here, i left a small comment as single comment instead of a full review that could be addressed. 

Additionally, ive noticed the caret-right, we're using from FA doesn't exactly match what we have in the mockup.

TASK-17
image

Mockup:
image

Im not sure if there is a similar one offered in FA, if not a small edit to the mockup should be done at some point.

Great work 👍

**Comment by cristian-aldea** [comment]
@JimLau49 
Yeah unfortuantely searching for icons I couldn't find one that matched the mockup. It doesn't really matter anyways

thanks, took a lot longer than I thought!

File: apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/event-page-header/event-page-header.tsx]: yup, removed the space

**Comment by cristian-aldea** [comment]
Alright I addressed all the comments I saw, lemme know if you see anything else.

**Review by JimLau49 — APPROVED** [review]
Review remains as an approve

**Review by RGPosadas — APPROVED** [review]

**Review by r-saba — APPROVED** [review]
LGTM! Great work

**Pull request merged**
